polytracker/cfg.py

```diff
@@ -0,0 +1,177 @@
+import math
+
+from typing import (
+    Any,
+    Callable,
+    Collection,
+    Dict,
+    FrozenSet,
+    Generic,
+    ItemsView,
+    Iterable,
+    KeysView,
+    List,
+    Optional,
+    Set,
+    TypeVar,
+)
+
+import graphviz
+import networkx as nx
+
+N = TypeVar("N")
+
+
+class DiGraph(nx.DiGraph, Generic[N]):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._dominator_forest: Optional[DiGraph[N]] = None
+        self._roots: Optional[Collection[N]] = None
+        self._path_lengths = None
+
+    def path_length(self, from_node, to_node):
+        if self._path_lengths is None:
+            self._path_lengths = dict(nx.all_pairs_shortest_path_length(self, cutoff=None))
+        if from_node not in self._path_lengths or to_node not in self._path_lengths[from_node]:
+            return math.inf
+        else:
+            return self._path_lengths[from_node][to_node]
+
+    def set_roots(self, roots: Collection[N]):
+        self._roots = roots
+
+    def _find_roots(self) -> Iterable[N]:
+        return (n for n, d in self.in_degree() if d == 0)
+
+    @property
+    def roots(self) -> Collection[N]:
+        if self._roots is None:
+            self._roots = tuple(self._find_roots())
+        return self._roots
+
+    def depth(self, node) -> int:
+        return min(self.path_length(root, node) for root in self.roots)
+
+    def ancestors(self, node) -> Set[N]:
+        return nx.ancestors(self, node)
+
+    def descendants(self, node) -> FrozenSet[N]:
+        return frozenset(nx.dfs_successors(self, node).keys())
+
+    @property
+    def dominator_forest(self) -> "DAG[N]":
+        if self._dominator_forest is not None:
+            return self._dominator_forest
+        self._dominator_forest = DAG()
+        for root in self.roots:
```

Comment:
Does this property access `self.roots` trigger the function above? (def roots)

Reply:
Correct!
```python
class Foo:
    _bar = 1337

    @property
    def bar(self):
        return self._bar

    @bar.setter
    def bar(self, value):
        self._bar = value

>>> foo = Foo()
>>> print(foo.bar)
1337
>>> foo.bar = 0
>>> print(foo.bar)
0
```
If you don't define a `@bar.setter`, then you will get a runtime error if you try and set it. This is useful if you want to have member variables that are effectively read only.